ruskeyverbs/views.py

```diff
@@ -26,19 +51,618 @@ def index(request):
+            filter=Q(
+                example__performanceperexample__user_id=request.user.pk)
+            )
+        ).order_by('due').exclude(due=None)
+    verbs_without_due_dates = Verb.objects.annotate(
+        due=Min('example__performanceperexample__due_date',
+                filter=Q(
+                    example__performanceperexample__user_id=request.user.pk)
+                )
+        ).exclude(~Q(due=None))
+    sorted_verb_list = list(chain(verbs_with_due_dates,
+                                  verbs_without_due_dates))
     return render(
         request,
         'ruskeyverbs/list_view_per_user.html',
-        context={'verb_list': sorted_verb_list, 'current_user': current_user}
+        context={'verb_list': sorted_verb_list}
     )
 
 
-class VerbDetails(LoginRequiredMixin, generic.DetailView):
-    model = Verb
+@login_required
+def VerbDetails(request, pk):
+    """
+    detail view for verbs; page will display conjugation and example info/audio
+    for user
+    """
+    verb = Verb.objects.get(pk=pk)
+    example_pk = Example.objects.annotate(
+        due=Min('performanceperexample__due_date',
+                filter=Q(
+                    performanceperexample__user_id=request.user.pk)
+                )
+        ).filter(verb=verb).order_by('due')[0].pk
+    """
+    If the user has previously studied the example, send them right
+    into the quiz; otherwise, send them to the study page before the
+    quiz
+    """
+    try:
+        print(PerformancePerExample.objects.get(pk=example_pk,
+                                                user=request.user))
+        not_studied = False
+    except ObjectDoesNotExist:
+        not_studied = True
+
+    request.session['quiz_counter'] = 0
+    # quiz_coutner will determine when the user is bumped out of the quiz loop
+    request.session['progress'] = 0
+    # progress is used to fill in the progress bar
+    request.session['quiz_summary'] = []
+    """quiz_summary will contain tuples of information on quiz performance
+    (Russian_Sentence, Translation_Text, Quiz_1_Score, Quiz_2_Score,
+    Quiz_3_score, Average_Score )"""
+    request.session['quiz_state'] = 0
+    # quiz state will restrict the user to going through the quizzes in order
+    return render(
+        request,
+        'ruskeyverbs/verb_detail.html',
+        context={'verb': verb,
+                 'example_pk': example_pk,
+                 'not_studied': not_studied}
+    )
+
+
+@login_required
+def StudyPage(request, pk):
+    """
+    page will be displayed before quiz if the user has not yet studied a verb
+    """
+    example_inst = get_object_or_404(Example, pk=pk)
+    russian_text_list = strip_punct_lower(example_inst.russian_text,
+                                          stressed=True).split()
+    """
+    test each word of the sentence to see which one is a form of the verb
+    being studied and determine which form it is
+    """
+    for i in range(len(russian_text_list)):
```

Comment:
Why not just for target_verb in russian_text_list:
...

It's much more Pythonic that way...

Reply:
Got rid of all for loops using range(len()).

Reply:
Good